File: Amia_dual.py
class dual_number(ob):
  real: float
  imag: float
  def __init__(self, real: float, imag: float):
    self.real = real
    self.imag = imag
  def con(self):
    z = dual_number(self.real, -self.imag)
    return z
  def toString(self):
    if self.imag >= 0:
        z = str(self.real) + '+' + str(self.imag) + 'ε'
        return z
     else:
        z = str(self.real) + str(self.imag) + 'ε';
        return z;
def add(x: dual_number, y: dual_number):
    real = x.real + y.real
    imag = x.imag + y.imag
    z = dual_number(real, imag)
    return z
def sub(x: dual_number, y: dual_number):
    real = x.real - y.real
    imag = x.imag - y.imag
    z = dual_number(real, imag)
    return z
def mul(x: dual_number, y: dual_number):
    real = x.real * y.real
    imag = x.real * y.imag + x.imag * y.real
    z = dual_number(real, imag)
    return z
def div(x: dual_number, y: dual_number):
    k = y.con()
    amia = dual_number(y.real * y.real, 0)
    z = mul(x, mul(k,amia))
    return z

File: TypeScript/Amia_dual/Amia_dual.ts
export class dual_number{
real: number
imag: number
    constructor(real: number, imag: number){
    this.real = real;
    this.imag = imag;
    }
    con(){
    const z: dual_number = new dual_number(this.real, -this.imag);
    return z;
    }
    toString(){
        if(this.imag >= 0){
            const z: string = this.real.toString() + '+' + this.imag.toString() + 'ε';
            return z;
        }
        else{
            const z: string = this.real.toString() + this.imag.toString() + 'ε';
            return z;
        }
    }
}
export function add(x: dual_number, y: dual_number): dual_number{
    const real = x.real + y.real;
    const imag = x.imag + y.imag;
    const z = new dual_number(real, imag);
    return z;
}
export function sub(x: dual_number, y: dual_number): dual_number{
    const real = x.real - y.real;
    const imag = x.imag - y.imag;
    const z = new dual_number(real, imag);
    return z;
}
export function mul(x: dual_number, y: dual_number): dual_number{
    const real = x.real * y.real;
    const imag = x.real * y.imag + x.imag * y.real;
    const z = new dual_number(real, imag);
    return z;
}
export function div(x: dual_number, y: dual_number): dual_number{
    const real = x.real/y.real;
    if(y.imag/y.real == 0 || x.real/y.real == 0){
        const imag = x.imag/y.real + (x.real * y.imag)/y.real;
    }else if(x.real * y.imag == Infinity || x.real * y.imag == -Infinity){
        if(x.real/y.real == 0){
            const imag = x.imag/y.real + x.real * (y.imag / y.real);
        }else{
            const imag = x.imag/y.real + y.imag * (x.real / y.real)
        }
    }
    const z = new dual_number(real,imag) ;
    return z;
}
